tower/src/retry/policy.rs

```diff
@@ -37,14 +38,14 @@ use std::future::Future;
 ///         }
 ///     }
 ///
-///     fn clone_request(&self, req: &Req) -> Option<Req> {
+///     fn clone_request(&mut self, req: &Req) -> Option<Req> {
 ///         Some(req.clone())
 ///     }
 /// }
 /// ```
-pub trait Policy<Req, Res, E>: Sized {
+pub trait Policy<Req, Res, E> {
     /// The [`Future`] type returned by [`Policy::retry`].
-    type Future: Future<Output = Self>;
+    type Future: Future<Output = ()>;
```

Comment:
changing this to return a `Future<Output = ()>` and duplicating the `Policy` using `Clone` rather than having the future return a new policy definitely simplifies things, but it occurs to me that asynchronous work can no longer be performed in order to update the `Policy`. i can't _immediately_ come up with a reason that it would be necessary to generate the next `Policy` asynchronously, but it seems like it could be worth thinking about before we commit to this design...?

Reply:
thinking about it, i suppose that if an implementation needed to modify the `Policy` after the future's completion, it could always clone an `Arc`ed shared state into its `Future`...this may introduce *some* overhead over the current approach, but I don't really think that use case is common enough that it's particularly important to worry about...

Reply:
Yeah, I don't think that use case is something we should support with that level and anyways taking a lock on an uncontested arc/mutex should be very cheap and much cheaper than request io etc.